README.md
# Simple HTTP Image Server
[![GitHub](https://img.shields.io/github/license/nikhilweee/shis)](https://github.com/nikhilweee/shis/blob/main/LICENSE.md)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/shis)](https://pypi.org/project/shis/)
[![PyPI](https://img.shields.io/pypi/v/shis)](https://pypi.org/project/shis/)
[![Documentation Status](https://readthedocs.org/projects/shis/badge/?version=stable)](https://shis.readthedocs.io)

A drop-in replacement for `python -m http.server`, albeit for images.


# Quickstart
Install SHIS.
```
pip install shis
```
Navigate to a `directory/containing/images`.
```
cd /directory/containing/images
```
Remember `python -m http.server`? Great. SHIS runs the same way.
```
python -m shis.server
```
There. You can now head over to http://0.0.0.0:7447/ (Or use your public IP instead). SHIS will generate thumbnails in the meanwhile.
```
# Serving HTTP on 0.0.0.0:7447. Press CTRL-C to quit.
# Processing images from : directory/containing/images
# Creating thumbnails in : directory/containing/images/shis
# Generating Website     : 100%|████████████████████|     4/    4 [00:00<00:00,  49.82it/s]
# Generating Thumbnails  : 100%|████████████████████|   300/  300 [00:00<00:00,  1.47kit/s]
```
**TIP**: You can install the latest development version directly from GitHub.
```
pip install git+https://github.com/nikhilweee/shis/
```


# Preview
<!--
    # shutil.copy(in_file, full_file) instead of os.symlink(full_dest, full_file)
    python -m shis.server -d imagenet-sample-images -s demo -n 100 -f -c
    find demo -type f -name "*.html" -exec sed -i "s/\"\//\"\/shis\//g" {} \;
    git subtree push --prefix demo/ origin gh-pages
-->
Here's an example of what you can expect to see. A live preview is also available at
[nikhilweee.github.io/shis](https://nikhilweee.github.io/shis).
![Demo](https://raw.githubusercontent.com/nikhilweee/shis/main/static/demo.png)


# Features
* Drop-in replacement for `python -m http.server`, so it's easy on your brain.
* Serves website even before creating thumbnails, so you don't have to wait.
* Uses multiple processes to create thumbails, so it's fast.
* Efficient resumes, so we build on past progress.
* Creates both small and large size thumbnails, so it's easy on your eyes.
* Minimal dependencies - Pillow, Jinja2, tqdm, and imagesize.
* Server side pagination, so it's easy on your browser.
* Tries to preserve EXIF orientation, so you don't have to rotate manually.
* Displays the public IP (if exists), so you don't have to remember.
* Watches the filesystem continuously for changes, so you don't have to refresh.
* Added selection capabilities, so you can visually filter files.


# Usage
The following options are available. You can also access this help using `python -m shis.server -h`. Further documentation can be found at [shis.readthedocs.io](https://shis.readthedocs.io).
```
usage: python -m shis.server [-h] [-c] [-s] [-p PORT] [-d DIR] [-w [SEC]]
                             [-n ITEMS] [-g ITEMS] [-o ORDER] [--thumb-dir DIR]
                             [--previews] [--ncpus CPUS] [--thumb-size SIZE]
                             [--preview-size SIZE]

A drop in replacement for python -m http.server, albeit for images.

optional arguments:
  -h, --help            show this help message and exit
  -c, --clean           remove existing --thumb-dir (if any) before processing
  -s, --selection       enable selection mode on the website
  -p PORT, --port PORT  port to host the server on (default: 7447)
  -d DIR, --image-dir DIR
                        directory to scan for images (default: current dir)
  -w [SEC], --watch [SEC]
                        filesystem watch interval in seconds (default: False)
  -n ITEMS, --pagination ITEMS
                        number of items to display per page (default: 200)
  -g ITEMS, --group ITEMS
                        number of items to group together (default: None)
  -o ORDER, --order ORDER
                        image listing order: name (default), random, or original
  --thumb-dir DIR       directory to store generated website (default: shis)
  --previews            also generate fullscreen previews (takes more time)
  --ncpus CPUS          number of workers to spawn (default: all available CPUs)
  --thumb-size SIZE     size of generated thumbnails in pixels (default: 256)
  --preview-size SIZE   size of generated previews in pixels (default: 1024)
```


# Benchmarks
For comparison, I ran the following tools on the [FFHQ Dataset](https://github.com/NVlabs/ffhq-dataset). The dataset contains 70k images in 1024x1024 resolution for a total size of 90GB. The converted thumbnail size was set to 320x320 for all tools. The tests were done on a machine with an AMD EPYC 7401P CPU with 24 Cores, 32GB Memory and Python 3.6.10 running on Ubuntu 18.04. The config files used are provided below. All conversion times are in `hh:mm:ss` format.

| Library Version | Conversion Time |             Configuration             |
|:---------------:|:---------------:|:-------------------------------------:|
|    shis 0.0.5   |      22:50      |                default                |
|   sigal 2.1.1   |      33:39      | [sigal.conf.py](static/sigal.conf.py) |
| thumbsup 2.14.0 |       >1h       | [thumbsup.json](static/thumbsup.json) |


# Why SHIS?
There are a bunch of static image servers (thumbsup, sigal, etc) available in a bunch of different languages (javascript, python, etc). While some of them like fgallery and curator haven't been developed in a while, others like thumbsup and sigal take a lot of time converting images. SHIS is designed with just one use case in mind, and it plans to do it well. It aims to serve a large directory of images in the fastest and easiest way possible.


# Acknowledgements
The demo at [nikhilweee.github.io/shis](https://nikhilweee.github.io/shis) shows sample images from the ImageNet dataset obtained through [imagenet-sample-images](https://github.com/EliSchwartz/imagenet-sample-images). The gallery template used to display images is a modified version of the cards theme from [thumbsup](https://github.com/thumbsup/theme-cards). SHIS also uses [lightgallery](https://github.com/sachinchoolur/lightGallery) for fullscreen previews and image slideshows.


# License
MIT License

CHANGELOG.md
# Changelog
All notable changes to [SHIS](https://github.com/nikhilweee/shis) will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Addded
- This CHANGELOG file to keep track of changes.
- A *watch* feature to continuously update the website based on filesystem changes.
- A *selection* mode to select multiple file names and copy them to the clipboard - useful for filtering images.
- Two icons on each image to open the image in gallery view and in new tab respectively.
- Numbering on each image, representing the index of the image in the folder.
- The ability to group images together, essentially serving as a visual aid.

### Changed
- Switched from `Pillow` to `imagesize` for faster determination if image sizes.
- SHIS will also clean up before exiting if the `-c` option is passed.
- Moved from absolute URLs to relative URLs in the generated site.
- Switched to dynamic versioning using `setuptools-git-versioning`

### Fixed
- A bug in determining the images that have changed and need to be processed again.
- Thumbnails are now resized according to the smallest dim so images with large aspect ratios don't appear blurry.
- A bug in determining public IPs in the first run.



## [0.1.0] - 2021-01-06
### Added
- The ability to automatically choose the next available port for starting a server.
- A [demo](https://nikhilweee.github.io/shis) hosted on GitHub pages.
- The ability to display the public IP of the machine if available.
- The ability to sort image files by `name`, or in `random` order.

### Changed
- The description bar format to handle large numbers.
- HTTP server protocol updated to HTTP/1.1.

### Fixed
- A bug in calculating the number of images that are to be processed.

[Unreleased]: https://github.com/nikhilweee/shis/compare/v0.1...HEAD
[0.1.0]: https://github.com/nikhilweee/shis/releases/tag/v0.1

shis/utils.py
import os
import sys
import argparse
import urllib.request
from functools import partial
from http.server import HTTPServer, SimpleHTTPRequestHandler
from threading import Thread
from typing import Generator, List, Tuple

from tqdm import tqdm


#-------------------------------------------------------------------------------
# General Utils
#-------------------------------------------------------------------------------


def chunks(iterable: List[str], chunk_size: int) -> Generator[List[str], None, None] :
    """Yield successive :attr:`chunk_size` sized chunks from :attr:`iterable`.

    :param iterable: an iterable to split into chunks.
    :param chunk_size: number of chunks to split :attr:`iterable` into.
    :return: a generator comtaining chunks of :attr:`iterable`.
    """
    for i in range(0, len(iterable), chunk_size):
        yield iterable[i:i + chunk_size]


def rreplace(string: str, find: str, replace: str) -> str:
    """Starting from the right, replace the first occurence of
    :attr:`find` in :attr:`string` with :attr:`replace`.

    :param string: the string to search :attr:`find` in.
    :param find: the substring to find in :attr:`string`.
    :param replace: the substring to replace :attr:`find` with.
    :return: the replaced string.
    """
    return replace.join(string.rsplit(find, 1))


def urlify(slug: str, page=1) -> str:
    """Create a URL given a :attr:`slug` and a :attr:`page` index.

    :param slug: a slug from :func:`slugify`.
    :param page: an optional page number to include in the url.
    :return: the path of the HTML page described by :attr:`slug`.
    """
    if page > 1:
        url = f'{slug}/page/{page}/'
    else:
        url = f'{slug}/'
    return url


def filter_image(name: str) -> bool:
    """Checks if a given file name is an image.

    :param name: the file name to check.
    :return: ``True`` if the file name is an image, ``False`` otherwise.
    """
    _, ext = os.path.splitext(name)
    if ext.lower() in ['.jpeg', '.jpg', '.png', '.tiff']:
        return True
    return False


def find_thumb(album_path: str, image_root: str, thumb_dir: str) -> str:
    """Finds the first available image in a directory.

    :param album_path: the directory to scan.
    :param image_root: the path to thumbnails of images in :attr:`album_path`.
    :param thumb_dir: the path to the generated website.
    :return: relative path to an image in :attr:`album_path` if it
        exists, empty string otherwise.
    """
    image = ''
    for root, _, files in os.walk(album_path):
        for name in files:
            file_path = os.path.join(root, name)
            if filter_image(file_path):
                rel_path = os.path.relpath(file_path, album_path)
                image_path = os.path.join(image_root, rel_path)
                image = os.path.relpath(image_path, thumb_dir)
                return image
    return image

def scale_dims(width: int, height: int, min_val: int) -> Tuple[int, int]:
    """Scales :attr:`width` and :attr:`height` according to :attr:`min_val`.

    The smaller out of :attr:`width` and :attr:`height` is assigned a value of
    :attr:`min_val`, and the other parameter is scaled accordingly.

    :param width: the width to scale
    :param height: the height to scale
    :param min_val: the minimum value of width or height
    :return: a tuple containing the scaled width and height
    """
    if width < height:
        width = width * min_val / height
        height = min_val
    if width == height:
        width = min_val
        height = min_val
    if width > height:
        width = width * min_val / height
        height = min_val
    return round(width), round(height)


#-------------------------------------------------------------------------------
# Argparse Utils
#-------------------------------------------------------------------------------


def fixed_width_formatter(width: int=80) -> argparse.HelpFormatter:
    """Patch :class:`argparse.HelpFormatter` to use a fixed width.

    :param width: the maximum width of the help and usage text generated.
    :return: a patched instance of the formatter class.
    """

    class HelpFormatter(argparse.HelpFormatter):

        def __init__(self, *args, **kwargs):
            super().__init__(width=width, *args, **kwargs)

    return HelpFormatter


#-------------------------------------------------------------------------------
# Server Utils
#-------------------------------------------------------------------------------


class CustomHTTPHandler(SimpleHTTPRequestHandler):
    """An HTTP Handler to serve arbitrary directories compatible with Python 3.6.

    This handler uses :attr:`self.server.directory` instead of always using
    ``os.getcwd()``

    :meta private:
    """

    protocol_version = "HTTP/1.1"

    def translate_path(self, path: str) -> str:
        """Translates a path to the local filename syntax."""
        path = SimpleHTTPRequestHandler.translate_path(self, path)
        if hasattr(self.server, 'directory'):
            relpath = os.path.relpath(path, os.getcwd())
            path = os.path.join(self.server.directory, relpath)
        return path
    
    def log_message(self, format: str, *args: str) -> None:
        """A dummy function overridden to disable logging."""
        pass
    
    def handle(self) -> None:
        """Handle multiple requests if necessary."""
        self.close_connection = True
        try:
            self.handle_one_request()
            while not self.close_connection:
                self.handle_one_request()
        except (ConnectionResetError, BrokenPipeError):
            pass


def start_server(args: argparse.Namespace) -> HTTPServer:
    """Start a Simple HTTP Server as a separate thread.
    
    :param args: preprocessed command line arguments.
    """
    # We need to create index.html for get_public_ip to receive HTTP 200.
    os.makedirs(args.thumb_dir, exist_ok=True)
    redir_html = os.path.join(args.thumb_dir, 'index.html')
    with open(redir_html, 'w') as f:
        f.write(f'<html><head><meta http-equiv="Refresh" '
                f'content="0; URL=html/"></head></html>')
    if sys.version_info.minor in [6, 7]:
        return start_server_36(args)
    if sys.version_info.minor >= 8:
        return start_server_38(args)


def get_public_ip(host: str, port: int) -> Tuple[str, int]:
    """Try to determine the public IP of the server.

    :param host: the fallback host to return in case of an error
    :param port: the port to check for public availability
    """
    try:
        with urllib.request.urlopen('https://api.ipify.org', timeout=5) as r:
            public_host = r.read().decode('utf-8')
            shis_server_url = f'http://{public_host}:{port}/'
        with urllib.request.urlopen(shis_server_url, timeout=5) as r:
            status = r.getcode()
        if status == 200:
            host = public_host
    except urllib.error.URLError:
        pass
    return host, port


def start_httpd(server: HTTPServer, address: Tuple[str, int], 
    handler: SimpleHTTPRequestHandler, args: argparse.Namespace) -> HTTPServer:
    """Try to start an HTTPServer, choosing the next available port.

    :param server: the server class to execute
    :param address: the address to start the server on
    :param handler: the request handler to use with the server
    :param args: preprocessed command line arguments
    """
    try:
        return server(address, handler)
    except OSError as error:
        if str(error) == '[Errno 98] Address already in use':
            if args.port is not None or address[1] > 7500:
                print(f'OSError: {error}. Try a different port using the -p flag.')
                sys.exit()
            else:
                address = (address[0], address[1]+1)
                return start_httpd(server, address, handler, args)
        else:
            raise error


def start_server_36(args):
    """Start an HTTP Server on Python 3.6 and Python 3.7.
    
    :meta private:
    """
    import socketserver

    class CustomHTTPServer(HTTPServer):
        def __init__(self, server_address: str, 
                     RequestHandlerClass: HTTPServer=CustomHTTPHandler,
                     directory: str=os.getcwd()):
            self.directory = directory
            HTTPServer.__init__(self, server_address, RequestHandlerClass)

    class ThreadingHTTPServer(socketserver.ThreadingMixIn, CustomHTTPServer):
        daemon_threads = False
    
    handler_class = CustomHTTPHandler
    server_class = partial(ThreadingHTTPServer, directory=args.thumb_dir)
    server_address = ("", args.port or 7447)
    httpd = start_httpd(server_class, server_address, handler_class, args)

    Thread(target=httpd.serve_forever).start()

    host, port = httpd.socket.getsockname()
    host, port = get_public_ip(host, port)
    serve_message = "Serving HTTP on {host}:{port}. "
    serve_message += "Press CTRL-C to quit."
    tqdm.write(serve_message.format(host=host, port=port))

    return httpd


def start_server_38(args):
    """Start an HTTP Server on Python 3.8 and above.
    
    :meta private:
    """    
    import contextlib
    from http.server import ThreadingHTTPServer, _get_best_family

    class DualStackServer(ThreadingHTTPServer):
        def server_bind(self):
            # suppress exception when protocol is IPv4
            with contextlib.suppress(Exception):
                self.socket.setsockopt(
                    socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
            return super().server_bind()

    handler_class = partial(CustomHTTPHandler, directory=args.thumb_dir)
    server_class = DualStackServer
    server_class.address_family, server_address = \
        _get_best_family(None, args.port or 7447)
    httpd = start_httpd(server_class, server_address, handler_class, args)
    
    Thread(target=httpd.serve_forever).start()

    host, port = httpd.socket.getsockname()[:2]
    host, port = get_public_ip(host, port)
    tqdm.write(f"Serving HTTP on {host}:{port}. "
               f"Press CTRL-C to quit.")

    return httpd


shis/server.py
import argparse
import math
import os
import sys
import shutil
import random
import time
from itertools import repeat
from multiprocessing import cpu_count
from typing import Dict, Tuple

import imagesize
from tqdm import tqdm
from PIL import Image, ImageOps
from tqdm.contrib.concurrent import process_map
from jinja2 import Environment, FileSystemLoader, select_autoescape

from shis.utils import (chunks, filter_image, find_thumb, rreplace, urlify,
                        start_server, scale_dims, fixed_width_formatter)


def generate_thumbnail(paths: Tuple[str, str, str, str], args: argparse.Namespace):
    """Takes paths from :func:`process_paths` and generates thumbnail(s).

    By default, only one thumbnail of size :attr:`args.thumb_size` is 
    created, while previews are symlinked to the original image.
    If :attr:`args.previews` is set, previews of :attr:`args.preview_size` 
    will also be created. Download links are always symlinked to the 
    original image.

    :param paths: A tuple of paths to process.
    :param args: preprocessed command line arguments.
    """

    in_file, small_file, large_file, full_file = paths
    if os.path.exists(small_file):
        if os.path.getmtime(small_file) >= os.path.getmtime(in_file):
            return
    try:
        # Save Preview
        im = Image.open(in_file)
        if args.previews:
            max_size = scale_dims(im.width, im.height, args.preview_size)
            im.thumbnail(max_size)
            im = ImageOps.exif_transpose(im)
            if 'exif' in im.info:
                exif = im.info['exif']
                im.save(large_file, exif=exif)
            else:
                im.save(large_file)
        # Save Thumbnail
        max_size = scale_dims(im.width, im.height, args.thumb_size)
        im.thumbnail(max_size)
        im = ImageOps.exif_transpose(im)
        if 'exif' in im.info:
            exif = im.info['exif']
            im.save(small_file, exif=exif)
        else:
            im.save(small_file)
        # Save Full
        full_dest = os.path.relpath(in_file, os.path.dirname(full_file))
        os.symlink(full_dest, full_file)
        return
    except Exception as e:
        return e


def process_paths(args: argparse.Namespace) -> Tuple[Tuple[str, str, str, str], int, bool]:
    """Generate paths to be processed by :func:`generate_thumbnail`

    If :attr:`args.clean` is set, all image files within :attr:`args.image_dir`
    will be included. Else, thumbnails which already exist and are newer than 
    the original image will be fileterd out.

    :param args: preprocessed command line arguments.
    :return: a tuple of (paths, num_pages, stale).

        - **paths** (*tuple*) - a 4-tuple containing (1) the absolute
          path of the original image, and the absolute path of the (2) small,
          (3) large and (4) full size thumbnails to be generated by
          :func:`generate_thumbnail`.
        - **num_pages** (*int*) - the number of webpages to generate.
        - **stale** (*bool*) - indicates whether a stale thumbnail was deleted.
    """
    paths = []
    num_pages = 0
    stale = False

    if not args.quiet:
        tqdm.write(f'Processing images from : {args.image_dir}')
        if args.clean and os.path.isdir(os.path.join(args.thumb_dir, 'small')):
            tqdm.write(f'Removing existing data : {args.thumb_dir}')
            shutil.rmtree(args.thumb_dir)
        tqdm.write(f'Creating thumbnails in : {args.thumb_dir}')

    for image_root, _, files in os.walk(args.image_dir):
        if args.thumb_dir in image_root:
            continue
        if image_root.count('/') > 100:
            raise ValueError(f'Too many subdirectories: {image_root}')
        small_root = rreplace(image_root, args.image_dir, f'{args.thumb_dir}/small')
        large_root = rreplace(image_root, args.image_dir, f'{args.thumb_dir}/large')
        full_root = rreplace(image_root, args.image_dir, f'{args.thumb_dir}/full')
        os.makedirs(small_root, exist_ok=True)
        os.makedirs(large_root, exist_ok=True)
        os.makedirs(full_root, exist_ok=True)
        num_pages += 1
        for idx, name in enumerate(filter(filter_image, files)):
            image_path = os.path.join(image_root, name)
            small_path = os.path.join(small_root, name)
            large_path = os.path.join(large_root, name)
            full_path = os.path.join(full_root, name)
            if idx != 0 and idx % args.pagination == 0:
                num_pages += 1
            if not os.path.exists(small_path):
                paths.append((image_path, small_path, large_path, full_path))
            elif os.path.getmtime(small_path) < os.path.getmtime(image_path):
                paths.append((image_path, small_path, large_path, full_path))
        # Make a list of thumbnails that have to be deleted
        thumb_files = filter(filter_image, os.listdir(small_root))
        stale_files = list(set(thumb_files) - set(files))
        stale = True if stale_files else stale
        for name in stale_files:
            for size_root in [small_root, large_root, full_root]:
                size_path = os.path.join(size_root, name)
                os.remove(size_path) if os.path.isfile(size_path) else None
    return paths, num_pages, stale


def generate_albums(args: argparse.Namespace) -> Tuple[Dict, int]:
    """Generate data required to populate Jinja2 templates.

    This function generates the correct names and URLs for all 
    thumbnails, albums, breadcrumbs and previews required for
    populating each HTML page of the website.

    :param args: preprocessed command line arguments.
    :return: a generator which yields data required to populate each page.
    """
    small_base = os.path.join(args.thumb_dir, 'small')
    image_head, image_tail = os.path.split(args.image_dir)

    for index_root, folders, files in os.walk(args.image_dir):
        if args.thumb_dir in index_root:
            continue
        small_root = rreplace(index_root, args.image_dir, small_base)
        full_root = rreplace(small_root, 'small', 'full')
        large_root = rreplace(small_root, 'small', 'large')
        slug_name = os.path.relpath(index_root, image_head)
        slug_path = rreplace(slug_name, image_tail, 'html')

        if not args.previews:
            large_root = full_root

        name = os.path.basename(slug_name)
        album = {'name': name}
        files = list(filter(filter_image, files))

        # Breadcrumbs
        crumbs = []
        crumb_root = ''
        for name, path in zip(slug_name.split(os.path.sep),
            slug_path.split(os.path.sep)):
            crumb_root = os.path.join(crumb_root, path)
            url = urlify(crumb_root)
            crumb = {'name': name, 'url': url}
            crumbs.append(crumb)
        album['crumbs'] = crumbs

        # Sort
        if args.order == 'name':
            files = sorted(files)
            folders = sorted(folders)
        if args.order == 'random':
            random.shuffle(files)
            random.shuffle(folders)

        # Albums
        albums = []
        for folder_name in folders:
            album_path = os.path.join(index_root, folder_name)
            if args.thumb_dir in album_path:
                continue
            album_size = len(os.listdir(album_path))
            image_root = os.path.join(small_root, folder_name)
            image = find_thumb(album_path, image_root, args.thumb_dir)

            album_slug_path = os.path.join(slug_path, folder_name)
            url = urlify(album_slug_path)
            folder = {'image': image, 'url': url,
                      'name': folder_name, 'size': album_size}
            albums.append(folder)
        album['albums'] = albums

        # Pagination
        num_pages = max(1, math.ceil(len(files) / args.pagination))
        pagination = []
        for page in range(1, num_pages + 1):
            url = urlify(slug_path, page)
            url = url.strip('/')
            page = {'page': page, 'url': url, 'current': ''}
            pagination.append(page)
        album['pagination'] = pagination
        album['group'] = args.group
        album['url'] = album['pagination'][0]['url']
        album['revpath'] = os.path.relpath('.', album['url'])
        album['selection'] = args.selection
        album['start_idx'] = 0

        # Images
        for page, chunk in enumerate(chunks(files, args.pagination)):
            thumbs = []
            for name in chunk:
                small_path = os.path.join(small_root, name)
                large_path = os.path.join(large_root, name)
                full_path = os.path.join(full_root, name)
                real_path = os.path.join(index_root, name)
                try:
                    width, height = imagesize.get(real_path)
                except ValueError:
                    continue
                width, height = scale_dims(width, height, args.thumb_size)
                small = os.path.relpath(small_path, args.thumb_dir)
                large = os.path.relpath(large_path, args.thumb_dir)
                full = os.path.relpath(full_path, args.thumb_dir)
                thumb = {'name': name, 'small': small, 'large': large,
                         'full': full, 'width': width, 'height': height}
                thumbs.append(thumb)
            album['thumbs'] = thumbs
            if page > 0:
                album['pagination'][page - 1]['current'] = None
                album['start_idx'] = page * args.pagination
            album['pagination'][page]['current'] = 'current'
            album['url'] = album['pagination'][page]['url']
            album['revpath'] = os.path.relpath('.', album['url'])
            yield album, page

        # Ensure that empty folders are not skipped
        if not files:
            yield album, 0


def create_templates(args: argparse.Namespace, num_pages: int) -> None:
    """Generate HTML files and corresponding directories for the website.

    This function creates ``static`` and ``html`` directories inside
    :attr:`args.thumb_dir`. If the directories already exist, they will be 
    cleared before continuing. All static content (JS/CSS) is stored 
    in :attr:`args.thumb_dir` ``/static`` and all HTML files (except 
    ``index.html``) are stored in :attr:`args.thumb_dir` ``/html``. 
    The data required to populate Jinja2 templates is obtained 
    from :func:`generate_albums`.

    :param args: preprocessed command line arguments.
    :param num_pages: number of HTML files (pages) to create.
    """
    # Copy JS/CSS
    template_dir = os.path.join(os.path.dirname(__file__), 'templates')
    static_src = os.path.join(template_dir, 'static')
    static_dest = os.path.join(args.thumb_dir, 'static')
    html_dir = os.path.join(args.thumb_dir, 'html')
    # Remove existing directories
    if os.path.exists(static_dest):
        shutil.rmtree(static_dest)
    if os.path.exists(html_dir):
        shutil.rmtree(html_dir)
    shutil.copytree(static_src, static_dest)
    os.makedirs(html_dir)
    # Generate fresh HTML
    env = Environment(
        loader=FileSystemLoader(template_dir),
        autoescape=select_autoescape(['html', 'xml'])
    )

    with tqdm(generate_albums(args),
        desc="Generating Website     ", total=num_pages, ncols=100,
        bar_format=("{l_bar}{bar:20}| {n_fmt:>5}/{total_fmt:>5} "
        "[{elapsed}<{remaining}, {rate_fmt:>10}{postfix}]")) as pbar:
        for album, page in generate_albums(args):
            template = env.get_template('index.html')
            url = album['pagination'][page]['url']
            html = f'{args.thumb_dir}/{url}/index.html'
            html_dir = os.path.dirname(html)
            if not os.path.isdir(html_dir):
                os.makedirs(html_dir, exist_ok=True)
            template.stream(album=album).dump(html)
            # Handle tqdm when files are still being added
            pbar.update(1)
            if pbar.n > pbar.total:
                pbar.total = pbar.n


def preprocess_args(args: argparse.Namespace) -> argparse.Namespace:
    """Preprocess arguments parsed by argparse.

    Specifically, this function disambiguates the relative
    paths supplied for :attr:`args.image_dir` and :attr:`args.thumb_dir`
    and converts them into absolute paths.

    :param args: command line arguments parsed by argparse.
    :return: preprocessed command line arguments.
    """
    args.image_dir = os.path.abspath(os.path.join(
        os.getcwd(), args.image_dir)).rstrip(os.path.sep)
    args.thumb_dir = os.path.abspath(os.path.join(
        os.getcwd(), args.thumb_dir)).rstrip(os.path.sep)
    if args.group:
        args.pagination += args.group - (args.pagination % args.group)
    return args


def main(args: argparse.Namespace) -> None:
    """The main script for coordinating shis.

    :param args: command line arguments parsed by argparse.
    """
    args = preprocess_args(args)
    # Start the server process
    try:
        server = start_server(args)
        stale_paths = []
        while True:
            # Generate HTML pages
            paths, num_pages, stale = process_paths(args)
            new_paths = list(set(paths) - set(stale_paths))
            if new_paths or stale:
                create_templates(args, num_pages)
            # Generate thumbnails
            if paths:
                process_map(generate_thumbnail, paths, repeat(args),
                    chunksize=1, max_workers=args.ncpus, unit_scale=True,
                    desc='Generating Thumbnails  ', ncols=100,
                    bar_format=("{l_bar}{bar:20}| {n_fmt:>5}/{total_fmt:>5} "
                    "[{elapsed}<{remaining}, {rate_fmt:>10}{postfix}]"))
            stale_paths = paths
            args.quiet = True
            if not args.watch:
                break
            else:
                time.sleep(args.watch)
        while True:
            # Loop until a KeyboardInterrupt is received
            time.sleep(1)
    except KeyboardInterrupt:
        print('\nKeyboard interrupt received, exiting.')
        server.shutdown()
        if args.clean and os.path.isdir(args.thumb_dir):
            tqdm.write(f'Removing existing data : {args.thumb_dir}')
            shutil.rmtree(args.thumb_dir)
        sys.exit()


def make_parser() -> argparse.ArgumentParser:
    """Creates a parser with the specified options.

    :return: a parser with the specified options.
    """

    parser = argparse.ArgumentParser(
        prog='python -m shis.server', 
        description='A drop in replacement for python -m http.server, albeit for images.',
        formatter_class=fixed_width_formatter(width=80))
    parser.add_argument('-c', '--clean', action='store_true',
        help='remove existing --thumb-dir (if any) before processing')
    parser.add_argument('-s', '--selection', action='store_true',
        help='enable selection mode on the website')
    parser.add_argument('-p', '--port', type=int, default=None,
        help='port to host the server on (default: 7447)')
    parser.add_argument('-d', '--image-dir', default='', metavar='DIR',
        help='directory to scan for images (default: current dir)')
    parser.add_argument('-w', '--watch', type=int, default=False, const=30, nargs='?',
        metavar='SEC', help='filesystem watch interval in seconds (default: %(default)s)')
    parser.add_argument('-n', '--pagination', type=int, default=200, metavar='ITEMS',
        help='number of items to display per page (default: %(default)s)')
    parser.add_argument('-g', '--group', type=int, default=None, metavar='ITEMS',
        help='number of items to group together (default: %(default)s)')
    parser.add_argument('-o', '--order', default='name', metavar='ORDER',
        choices = ['original', 'random', 'name'],
        help='image listing order: name (default), random, or original')
    parser.add_argument('--thumb-dir', default='shis', metavar='DIR',
        help='directory to store generated website (default: %(default)s)')
    parser.add_argument('--previews', action='store_true',
        help='also generate fullscreen previews (takes more time)')
    parser.add_argument('--ncpus', type=int, default=cpu_count(), metavar='CPUS',
        help='number of workers to spawn (default: all available CPUs)')
    parser.add_argument('--thumb-size', type=int, default=256, metavar='SIZE',
        help='size of generated thumbnails in pixels (default: %(default)s)')
    parser.add_argument('--preview-size', type=int, default=1024, metavar='SIZE',
        help='size of generated previews in pixels (default: %(default)s)')
    return parser


if __name__ == '__main__':
    parser = make_parser()
    args = parser.parse_args()
    args.quiet = False
    main(args)
